File: BFS/130. 被围绕的区域.py
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: None Do not return anything, modify board in-place instead.
        """

        class point():
            def __init__(self):
                self.x = None
                self.y= None
            def tostrint(self):
                print(self.x, self.y)

        high = len(board)
        width = len(board[0])

        queue = []

        il =[0, high-1]
        for j in range(width):
            for i in il:
                if board[i][j] == 'O':
                    p = point()
                    p.x=j
                    p.y=i

                    queue.append(p)
        jl = [0, width-1]
        for i in range(1, high-1):
            for j in jl:
                if board[i][j] == 'O':
                    p = point()
                    p.x = j
                    p.y = i

                    queue.append(p)
        for i in queue:
            i.tostrint()

        #以上完成了基本得边界检测
        import copy
        boardNew = copy.deepcopy(board)

        #代表是否来过
        res=[]
        for i in range(high):
            tmp =[]
            for j in range(width):
                tmp.append(False)
                boardNew[i][j] = 'X'
            res.append(tmp)
        # print(res)


        while len(queue) !=0:

            top = queue.pop()
            i = top.y
            j = top.x
            boardNew[i][j] = 'O'
            res[i][j] = True

            il = [1, -1]
            jl = [1, -1]
            #向上下左右四个方向进行遍历
            for m in il:
                for n in jl:
                    tmpi = i + m
                    tmpj = j + n

                    if tmpi <0 or tmpi >= high or tmpj <0 or tmpj >= width:
                        continue

                    if res[tmpi][tmpj] and (board[tmpj][tmpj] == 'O'):
                        p = point()
                        p.x = tmpj
                        p.y = tmpi
                        queue.append(p)
        print(boardNew)
        return  boardNew










Solution.solve(None,[["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]])

File: 图/other/130. 被围绕的区域.java
class Solution {
    public void solve(char[][] board) {
        // 从边界开始搜索
        int m = board.length;
        int n = board[0].length;
        for(int i = 0; i < m; i++){
            dfs(board, i, 0);
            dfs(board, i, n-1);
        }
        for(int i = 0; i < n; i++){
            dfs(board, 0, i);
            dfs(board, m-1, i);
        }

        // 将所有的#变成O，所有的O变成X
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
                else if (board[i][j] == '#') {
                    board[i][j] = 'O';
                }
            }
        }
    }

    public void dfs(char[][] board, int x, int y){
        int m = board.length;
        int n = board[0].length;

        
        if (x >= 0 && x < m && y >= 0 && y < n && board[x][y] == 'O') {


            board[x][y] = '#';
            dfs(board, x, y-1);
            dfs(board, x, y+1);
            dfs(board, x+1, y);
            dfs(board, x-1, y);
        }
    }
}